### PR TITLE
IEP-955 Refactoring SerialFlashLaunchConfigDelegate class

### PR DESCRIPTION
**Pull request by sigmaaa**
## Description

summary:
* Added openocd_path, openocd_exe, OPENOCD_SCRIPTS variable and resolver for these variables
* Moved responsibility of parsing dynamic variables from SerialFlashLaunchConfigDelegate to the variable resolver
* Added openocd prefix to the arguments line in the CMakeMainTab2 so the default value looks like this:
`${openocd_path}\${openocd_exe} -s ${OPENOCD_SCRIPTS} -f board/esp32-wrover-kit-1.8v.cfg`
* Added argument line check method in the SerialFlashLaunchConfigDelegate to provide backward compatibility for the use case, when `${openocd_path}\${openocd_exe}` prefix is missing from the argument line (Previous behavior) 
* Moved flash over jtag to the ESPFlashUtil class 
* Added tests for variable resolver

Fixes # ([IEP-955](https://jira.espressif.com:8443/browse/IEP-955))

## Type of change

Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How has this been tested?
Test 1:
- create a new project 
- modify launch configuration -> select `flash over JTAG`
- make sure that now the arguments line looks like this `${openocd_path}\${openocd_exe} -s ${OPENOCD_SCRIPTS} -f board/esp32-wrover-kit-1.8v.cfg`

Test 2:
- try flash over jtag with the default argument line
Test 3:
- try flash over jtag with modified argument line (remove ${openocd_path}\${openocd_exe} part) - should be working and proving backward compatibility 

**Test Configuration**:
* ESP-IDF Version:
* OS (Windows,Linux and macOS):

## Dependent components impacted by this PR:

- Argument line of the launch configuration with flash over jtag option
- flashing

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Added Documentation
- [x] Added Unit Test
- [ ] Verified on all platforms - Windows, Linux and macOS

### CODE REVIEW

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdDynamicVariable.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+	OpenocdDynamicVariable()
+	{
+		this.value = name();
+	}
+
+	public String getValue()
+	{
+		return value;
+	}
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdDynamicVariable.java:22]: could be simplified in the future

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdVariableResolver.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+	public String resolveValue(IDynamicVariable variable, String argument) throws CoreException
+	{
+		return getAppropriateEnumVariable(variable).map(this::resolveForOpenocdDynamicEnum).orElse(variable.getName());
+	}
+
+	private Optional<OpenocdDynamicVariable> getAppropriateEnumVariable(IDynamicVariable variable)
+	{
+		return Arrays.stream(OpenocdDynamicVariable.values()).filter(v -> v.getValue().equals(variable.getName()))
+				.findAny();
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdVariableResolver.java:34]: added this to make code more robust, consider this use case:
- adding the variable to the extension point, but not adding it to the openocd enum. 
- In this case, returning striped value without resolving it;

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdVariableResolver.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+		case OPENOCD_EXE:
+			return File.separator + "bin" + File.separator + "openocd"; //$NON-NLS-1$ //$NON-NLS-2$
+
+		case OPENOCD_SCRIPTS:
+			return new IDFEnvironmentVariables().getEnvValue(IDFEnvironmentVariables.OPENOCD_SCRIPTS);
+
+		default:
+			return StringUtil.EMPTY;
+		}
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/variable/OpenocdVariableResolver.java:54]: the default can be removed after migration to Java 14 or higher

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/ESPFlashUtil.java [hunk 4]
```diff
@@ -30,6 +37,8 @@
 	private static final int OPENOCD_JTAG_FLASH_SUPPORT_V = 20201125;
 	public static final String VERSION_PATTERN = "(v.\\S+)"; //$NON-NLS-1$
 	public static final String SERIAL_PORT = "${serial_port}"; //$NON-NLS-1$
+	// prefix for backward compatibility with 2.9.1 where this prefix was not added in the argument in the UI
+	private static final String DEFAULT_ARGUMENT_PREFIX = "${openocd_path}/${openocd_executable} "; //$NON-NLS-1$
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/ESPFlashUtil.java:41]: theoretically can be removed in the major release

**Review by sigmaaa — COMMENTED** [review]
Self reviewed

**Comment by sigmaaa** [comment]
added test coverage for variable resolver

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java:1]: In this class, we currently handle the 'serialPort' variable. However, we can introduce a dynamic variable for the serial port in the future, which will allow us to remove 'serialPort'. Additionally, we can enhance the user experience with the following improvement:

When attempting to flash with an empty serial port, a dialog will appear suggesting the user to select a serial port. If the user chooses to proceed, the flashing process will not terminate but instead, wait until a port is selected. Once a port is selected, the flashing will resume.

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 5]
```diff
@@ -792,6 +794,10 @@ private void updateArgumentsField() {
 		jtagArgumentsField.setText(configOptiopns);
 	}
 
+	private String newVariableExpression(OpenocdDynamicVariable dynamicVariable) {
+		return newVariableExpression(dynamicVariable.getValue(), null);
+	}
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:799]: I was thinking about moving this to OpenocdDynamicVariable responsibility, but I dropped this idea as there is little benefit as it is only used in this class. Also, this class already doing similar things with other variables.
In the future, we will probably add a similar thing to the TabMain class

**Review by sigmaaa — COMMENTED** [review]
self reviewed changes

File: bundles/com.espressif.idf.core/plugin.xml [hunk 6]
```diff
@@ -260,4 +260,22 @@ config-only component and an interface library is created instead."
...
+     </variable>
+     <variable
+           description="%openocd_exe"
+           name="openocd_exe"
+           resolver="com.espressif.idf.core.variable.OpenocdVariableResolver">
+     </variable>
+     <variable
+           description="%openocd_scripts"
+           name="OPENOCD_SCRIPTS"
+           resolver="com.espressif.idf.core.DynamicVariableResolver5">
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.core/plugin.xml:278]: DynamicVariableResolver5 ? I couldn't find relevant class for this?

**Review by kolipakakondal — APPROVED** [review]
I see byte-buddy.jar and byte-buddy-agent.jar and objenesis jars in the libs? If we are not using them we can remove them

**Comment by kolipakakondal** [comment]
Overall changes LGTM. @AndriiFilippov please go ahead with the review.

File: bundles/com.espressif.idf.core/plugin.xml (hunk 6 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.core/plugin.xml]: Hi @kolipakakondal, thanks for noticing. Looks like Eclipse didn't pick up this class after Refactoring -> Rename. Fixed in the latest commit

**Comment by sigmaaa** [comment]
> I see byte-buddy.jar and byte-buddy-agent.jar and objenesis jars in the libs? If we are not using them we can remove them

Hi @kolipakakondal, these are dependencies for Mockito

**Comment by AndriiFilippov** [comment]
@sigmaaa  hi !

Tested under: 
OS - WIndows 10 

**Test 3:**
try flash over jtag with modified argument line (remove ${openocd_path}${openocd_exe} part) - should be working and proving backward compatibility

Get this error: 

image

**Comment by sigmaaa** [comment]
> @sigmaaa hi !
> 
> Tested under: OS - WIndows 10
> 
> **Test 3:** try flash over jtag with modified argument line (remove ${openocd_path}${openocd_exe} part) - should be working and proving backward compatibility
> 
> Get this error: An exception stack trace is not available. image

Hi @AndriiFilippov, thank you for verifying. Must be fixed after last commit

**Comment by AndriiFilippov** [comment]
@sigmaaa hi ! 


LGTM 👍

**Pull request merged**